File: 강의 자료/02-알고리즘/07~08-두 포인터/문제별 코드/2230-수 고르기/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 100005

int N, M;
int A[NM];

void input() {
	cin >> N >> M;
	for (int i = 1; i <= N; i++) cin >> A[i];
}

void pro() {
	sort(A + 1, A + 1 + N);

	int R = 1, ans = 0x7fffffff;
	for (int L = 1; L <= N; L++) {
		while (R + 1 <= N && A[R] - A[L] < M) ++R;
		if (A[R] - A[L] >= M) ans = min(ans , A[R] - A[L]);
	}

	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/07~08-두 포인터/문제별 코드/2230-수 고르기/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static StringBuilder sb = new StringBuilder();
    static FastReader scan = new FastReader();

    static int n, M;
    static int[] a;

    static void input() {
        n = scan.nextInt();
        M = scan.nextInt();
        a = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            a[i] = scan.nextInt();
        }
    }

    static void pro() {
        // 정렬 해주기
        Arrays.sort(a, 1, n + 1);

        int R = 1, ans = Integer.MAX_VALUE;
        for (int L = 1; L <= n; L++) {
            // 필요한 만큼 R을 오른쪽으로 이동 시키기
            while (R + 1 <= n && a[R] - a[L] < M) ++R;

            // 정답 갱신하기
            if (a[R] - a[L] >= M) ans = Math.min(ans, a[R] - a[L]);
        }

        System.out.println(ans);
    }

    public static void main(String[] args) {
        input();
        pro();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}